lms/static/scripts/frontend_apps/services/client-rpc.ts

```diff
@@ -137,6 +147,28 @@ export class ClientRPC extends TinyEmitter {
       },
     );
 
+    // Listen for client notifications about unsaved changes.
+    //
+    // This is used to work around a restriction in desktop Safari where
+    // cross-origin iframes cannot use `beforeunload` to prevent tabs being
+    // closed. If the LMS frontend is the top-level frame, then we can work
+    // around this by having the client notify the LMS frontend, which then
+    // adds a "beforeunload" listener.
+    let clientHasUnsavedChanges = false;
+    this._server.register(
+      'reportUnsavedChanges',
+      (state: UnsavedChangesData) => {
+        if (clientHasUnsavedChanges !== state.unsaved) {
+          clientHasUnsavedChanges = state.unsaved;
+          if (state.unsaved) {
+            incrementUnsavedCount();
+          } else {
+            decrementUnsavedCount();
+          }
+        }
```

Comment:
Perhaps we can reduce nesting a bit.


```suggestion
        if (clientHasUnsavedChanges === state.unsaved) {
          return;
        }
        
        clientHasUnsavedChanges = state.unsaved;
        if (state.unsaved) {
          incrementUnsavedCount();
        } else {
          decrementUnsavedCount();
        }
```